# Datetime Module in Python

Python has datetime module to work with dates and times. The Python datetime module offers functions and classes for working with date and time parsing, formatting, and arithmetic. Let's have a look at some basic operations.

In [7]:
#Get Current Date and Time
import datetime
datetime_obj = datetime.datetime.now()
print(datetime_obj)

#Get Current Date
date_obj = datetime.date.today()
print(date_obj)

2019-07-04 20:47:26.831629
2019-07-04


As always, we imported datetime module using "import datetime" statement.

Datetime module has a few important classes. We will look at them as we move further. Here, we use the datetime class. We then used now() method in datetime class(of datetime module) to create a datetime object containing the current date and time.

In the next example, we have used today() method in the date class to get a date object containing the local date.

Now We will look individually at each component of the datetime module. 

The datetime module has two constants:

In [8]:
print(datetime.MINYEAR)
print(datetime.MAXYEAR)

1
9999


datetime.MINYEAR
The smallest year number allowed in a date or datetime object. MINYEAR is 1.

datetime.MAXYEAR
The largest year number allowed in a date or datetime object. MAXYEAR is 9999.

We have six classes in datetime module:
    1. class datetime.date
    2. class datetime.time
    3. class datetime.datetime
    4. class datetime.timedelta
    5. class datetime.tzinfo
    6. class datetime.timezone
    
We will look at the top four classes in detail. These are the most used classes. After that, we will specifically look at few important operations like conversion to different formats and datetime formatting. 

## Class datetime.date 

You can initialize date objects from the date class. A date object represents a date (year, month and day). The object takes three input numbers, i.e year, month and date.
date_obj = class datetime.date(year, month, day)
    MINYEAR <= year <= MAXYEAR
    1 <= month <= 12
    1 <= day <= number of days in the given month and year
    
ValueError is raised if the inputs are outside the specified ranges.

In [9]:
#Date object to represent a date
dt = datetime.date(2019, 4, 13)
print(dt)

2019-04-13


### Class Methods

In [10]:
#today method
from datetime import date
today = date.today()
print("Current date =", today)
print("Current date =", date.today())
print('\n')
#timestamp method
timestamp = date.fromtimestamp(1326244364)
print("Date =", timestamp)
print("Date =", date.fromtimestamp(1326244364))

Current date = 2019-07-04
Current date = 2019-07-04


Date = 2012-01-11
Date = 2012-01-11


### Instance Methods

Next, we look at some important instance attributes. Instance attributes work only on date objects created correspondng to date class.

date.replace(year, month, day) : Return a date with the same value, except for those parameters given new values by  keyword arguments inside braces. 

date.weekday() : Return the day of the week as an integer. starting form Monday = 0 to Sunday = 6.

date.isoweekday() : Return the day of the week as an integer, where Monday is 1 and Sunday is 7. Based on ISO calendar(https://www.staff.science.uu.nl/~gent0113/calendar/isocalendar.htm) 

date.isocalendar() : Return a 3-tuple (ISO year, ISO week number, ISO weekday).

date.isoformat() : Return a string representing the date in ISO 8601 format, ‘YYYY-MM-DD’.

date.ctime() : Return a string representing the date, e.g. date(2002, 12, 4).ctime() = 'Wed Dec 4 00:00:00 2002'. 

-------------------------New addition --------------------------------------------------
Note  : date.ctime()(and datetime.ctime()) outputs the corresponding date in a particular string format given above. We have another module called Time(not to be confused with datetime module's time class) which deals with complex date and time operations, which has the same ".ctime()" method, but in this case, .ctime(secs) takes input as seconds from epoch and returns the corresponding date and time in the format "".
-------------------------New addition --------------------------------------------------


date.strftime(format) : Return a date string in the given format, controlled by a defined string template. (See https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior for details on formatting and placeholders)


In [11]:
### Replace method
date_obj = date(2017, 6, 14)
print(date_obj)

d2 = date_obj.replace(day = 30)
print(d2)

### weekday, isoweekday, isocalendar method
print("weekday = " + str(d2.weekday()))
print("Iso weekday = " + str(d2.isoweekday()))
print("ISO calendar = {d}".format(d = d2.isocalendar()))
print("ISO format = {d}".format(d = d2.isoformat()))
print("date string = {d}".format(d = d2.ctime()))
print("date string in dd/mm/yy format = {d}".format(d = d2.strftime('%d/%m/%y')))

2017-06-14
2017-06-30
weekday = 4
Iso weekday = 5
ISO calendar = (2017, 26, 5)
ISO format = 2017-06-30
date string = Fri Jun 30 00:00:00 2017
date string in dd/mm/yy format = 30/06/17


### Attributes

Date class attributes:
Note that for class attributes, we need not necessarily define a date object. Look at the example below for better understanding. 

date.min : The earliest representable date i.e. the object corresponding to date(MINYEAR, 1, 1).

date.max : The latest representable date i.e. the object corresponding to date(MAXYEAR, 12, 31).

date.resolution : The smallest possible difference between two different dates, i.e. the object corresponding to timedelta(days=1).

read-only instance attributes:

date.year : any output between MINYEAR and MAXYEAR, it gives the year of corresponding date object.

date.month : any output between 1 and 12 including both, it gives the month of corresponding date object.

date.day : any output between 1 and the number of days in the given month of the given year, all data from the given date object.

In [12]:
from datetime import date

# create a date object named tdy
tdy = date.today()

# class attributes, work with or without defining the date object
print(tdy.min)
print(date.min)
print('\n')
# class attributes, work with or without defining the date object
print(tdy.max)
print(date.max)
print('\n')
#read-only instance attributes
print(tdy.year)
print(tdy.month)
print(tdy.day)

0001-01-01
0001-01-01


9999-12-31
9999-12-31


2019
7
4


## Class datetime.time

A time object represents a particular local time of day, independent of any particular day and can be adjusted by a tzinfo object corresponding to tzinfo class.

time_obj = datetime.time(hour=0, minute=0, second=0, microsecond=0, tzinfo=None, fold =0)

0 <= hour < 24,
0 <= minute < 60,
0 <= second < 60,
0 <= microsecond < 1000000,
fold in [0, 1].

In [5]:
from datetime import time

# time(hour = 0, minute = 0, second = 0)
a = time()
print("t0 =", a)
# time(hour, minute and second)
b = time(11, 34, 56)
print("t1 =", b)
# time(hour, minute and second)
c = time(hour = 11, minute = 34, second = 56)
print("t2 =", c)
# time(hour, minute, second, microsecond)
d = time(11, 34, 56, 234566)
print("t3 =", d)

t0 = 00:00:00
t1 = 11:34:56
t2 = 11:34:56
t3 = 11:34:56.234566


### Class Methods

time.fromisoformat('time_string') : Returns time corresponding to a string of time as input. the input can be in any of the string formats given by time.isoformat().

In [7]:
dt = time(hour=12, minute=34, second=56, microsecond=0)

print(time.fromisoformat(dt.isoformat(timespec='microseconds')))
print(time.fromisoformat(dt.isoformat(timespec='auto')))


12:34:56
12:34:56


### Instance Methods

time.isoformat(timespec='auto') : This returns a string representing the time in ISO 8601 format.

Timespec specifies the components of the time to which is to be included. 

    'auto': (default)if microsecond is 0, the format is similar to 'seconds', otherwise similar to 'microseconds'.
    'hours': Hour in the two-digit HH format.
    'minutes': Add hour and minute in HH:MM format.
    'seconds': Add hour, minute, and second in HH:MM:SS format.
    'milliseconds': Include full time, but fractional second part is to milliseconds. HH:MM:SS.sss format.
    'microseconds': Include full time in HH:MM:SS.ffffff format.
    
time.utcoffset() : A time offset is an amount of time subtracted from or added to Coordinated Universal Time (UTC) time to get the current time. If tzinfo is None, this returns None, else returns the time offset(a timedelta object with magnitude less than one day).

time.tzname() : If tzinfo is None, returns None, else returns the timezone name.

Instance methods like replace() and strftime() work similar to their date class counterparts.

In [15]:
dt = time(hour=12, minute=34, second=56, microsecond=0)

print(dt.utcoffset())
print(dt.isoformat(timespec='seconds'))
print(dt.tzname())

None
12:34:56
None


### Attributes

Class attributes:

time.min : The earliest representable time i.e. time(0, 0, 0, 0).

time.max : The last representable time i.e. time(23, 59, 59, 999999).

time.resolution : The smallest possible difference between non-equal time objects i.e. timedelta(microseconds=1).

Read-only Instance attributes :

time.hour : Output the hour component of time in range(0,24)

time.minute : Output the minutes component in range(0,60).

time.second : Output the seconds component in range(0, 60).

time.microsecond : Output the microseconds component in range(0,1000000).

time.tzinfo : Outputs the object passed as the tzinfo argument(which gives time zone information) or None if none was passed.

time.fold : It is used to differenciate similar wall times during a repeated interval. (A repeated interval occurs when clocks are rolled back at the end of daylight saving time(dst) or when the UTC offset for the current zone is decreased for political reasons.) The value 0 represents the earlier and 1 represents later of the two moments with the same wall time representation.

In [16]:
dt = time(hour=12, minute=34, second=56, microsecond=0)
#class attributes
print(time.min)
print(time.max)
print(time.resolution)
print('\n')
#instance Attributes
print(dt.hour)
print(dt.second)
print(dt.fold)

00:00:00
23:59:59.999999
0:00:00.000001


12
56
0


## Class datetime.datetime

The datetime module has a class named dateclass that can contain information from both date and time objects.

class datetime.datetime(year, month, day, hour=0, minute=0, second=0, microsecond=0, tzinfo=None, fold=0)

In [17]:
#Python datetime object
from datetime import datetime
#datetime(year, month, day)
a = datetime(2018, 11, 28)
print(a)
# datetime(year, month, day, hour, minute, second, microsecond)
b = datetime(2017, 11, 28, 23, 55, 59, 342380)
print(b)

2018-11-28 00:00:00
2017-11-28 23:55:59.342380


### Class Methods

We have class methods similar to the date and time classes in datetime classes as well. These include datetime.today(), datetime.fromtimestamp(timestamp, tz=None) and datetime.fromisoformat(date_string). Some new class methods are discussed below:

datetime.combine(date, time) : Return a new datetime object whose date components are equal to the given date object’s, and whose time components are equal to the given time object’s.

datetime.now(tz=None) : Return the current local date and time. If optional argument tz is None or not specified, this is like today()

datetime.strptime(date_string, format) : Return a datetime object corresponding to a date string, parsed according to the given datetime format.

In [18]:
#datetime.now() method
print(datetime.now())

#datetime.combine() method
a = datetime(2018, 11, 28)
b = datetime(2017, 11, 28, 23, 55, 59, 342380)
print(datetime.combine(a.date(), b.time()))

#datetime.strptime method
dt=datetime.strptime("21/11/06 16:30", "%d/%m/%y %H:%M")
print(dt)
dt

2019-07-04 20:47:55.899245
2018-11-28 23:55:59.342380
2006-11-21 16:30:00


datetime.datetime(2006, 11, 21, 16, 30)

### Instance Methods

We have datetime.date(), datetime.time(), datetime.replace(), datetime.tzname(), datetime.utcoffset(), datetime.weekday(), datetime.isoweekday(), datetime.isoformat(), datetime.ctime(), datetime.strftime(), datetime.timestampetc instance methods similar to the ones seen in date and time classes.

### Attributes

Class attributes : We have the similar class attributes as in previous classes i.e, datetime.max, datetime.min and datetime.resolution which work similar to previous cases.

Read-only attributes : we have attributes like datetime.year, month, day, hour, minute, second, microsecond, tzinfo and flod, all of which are obvious and are covered in previous sections  

## Class datetime.timedelta

A timedelta object represents a duration, the difference between two dates or times.

class datetime.timedelta(days, seconds, microseconds, milliseconds, minutes, hours, weeks)
All arguments are optional and default to 0. Arguments may be integers or floats, and may be positive or negative.

Only days, seconds and microseconds are stored internally. Rest arguments are converted to those units.

Please note that, although while defining, the timedelta object takes in optional inputs as days, seconds, microseconds, milliseconds, minutes, hours and weeks, it only stores the data in microseconds, seconds and days. So weeks are converted to days and added to "days" input argument, hours and minutes are converted to seconds and added to the "seconds" argument given as input and milliseconds are converted to microseconds and added the "microseconds" input argument. 

0 <= microseconds < 1000000

0 <= seconds < 3600*24 (the number of seconds in one day)

-999999999 <= days <= 999999999


In [10]:
from datetime import timedelta

t1 = timedelta(weeks = 3, days = 2, hours = 6, seconds = 23, milliseconds = 100)
print(t1)

##default time format is printed, i.e. milliseconds are printed in microseconds.

t2 = timedelta(weeks = 5, days = 41, hours = 1, minutes = 44, seconds = 34)
print(t2)

23 days, 6:00:23.100000
76 days, 1:44:34


In [11]:
#Instance method timedelta.total_seconds()
#Return the total number of seconds contained in the duration.

t2 = timedelta(weeks = 5, days = 41, hours = 1, minutes = 44, seconds = 34)
t2.total_seconds()

6572674.0

### Attributes

We have similar class attributes as in previous classes, i.e. min, max and resolution.

We have days, seconds and microseconds as instance attributes. Please note that we don't have hours or minutes or weeks as instance attributes as they are all converted to days, seconds and microseconds while storing.

In [12]:
t2 = timedelta(weeks = 3, days = 2, hours = 6, seconds = 23, milliseconds = 100)

#instance attributes
print(t2.days)
print(t2.seconds)
print(t2.microseconds)

#this shows an error
print(t2.hours)

23
21623
100000


AttributeError: 'datetime.timedelta' object has no attribute 'hours'

## References

https://www.programiz.com/python-programming/datetime#timedelta

https://docs.python.org/3/library/datetime.html

https://howchoo.com/g/ywi5m2vkodk/working-with-datetime-objects-and-timezones-in-python

https://www.w3schools.com/python/python_datetime.asp
